src/poisson_solvers.jl

```diff
@@ -20,127 +24,112 @@ let pf2s = Dict(FFTW.ESTIMATE   => "FFTW.ESTIMATE",
     plannerflag2string(k::Integer) = pf2s[Int(k)]
 end
 
-"""
-    PoissonSolver(grid, example_field, planner_flag; verbose=false)
+function PoissonSolverCPU(grid::Grid, planner_flag=FFTW.PATIENT)
+    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
+    Lx, Ly, Lz = grid.Lx, grid.Ly, grid.Lz
 
-Return a `PoissonSolver` on `grid`, using `example_field` and `planner_flag`
-to plan fast transforms.
-"""
-struct PoissonSolver{T<:AbstractArray} <: AbstractPoissonSolver
-    kx²::T
-    ky²::T
-    kz²::T
-    FFT!
-    DCT!
-    IFFT!
-    IDCT!
-end
+    # Storage for RHS and Fourier coefficients is hard-coded to be Float64 because of precision issues with Float32.
+    # See https://github.com/climate-machine/Oceananigans.jl/issues/55
+    kx² = zeros(Float64, Nx)
+    ky² = zeros(Float64, Ny)
+    kz² = zeros(Float64, Nz)
 
-function PoissonSolver(g::Grid, exfield::CellField, planner_flag=FFTW.PATIENT; verbose=false)
-    kx² = zeros(eltype(g), g.Nx)
-    ky² = zeros(eltype(g), g.Ny)
-    kz² = zeros(eltype(g), g.Nz)
-
-    for i in 1:g.Nx; kx²[i] = (2sin((i-1)*π/g.Nx)    / (g.Lx/g.Nx))^2; end
-    for j in 1:g.Ny; ky²[j] = (2sin((j-1)*π/g.Ny)    / (g.Ly/g.Ny))^2; end
-    for k in 1:g.Nz; kz²[k] = (2sin((k-1)*π/(2g.Nz)) / (g.Lz/g.Nz))^2; end
-
-    if verbose
-        print("Planning Fourier transforms... (planner_flag=$(plannerflag2string(planner_flag)))\n")
-        print("FFT!:  "); @time FFT!  = FFTW.plan_fft!(exfield.data, [1, 2]; flags=planner_flag)
-        print("IFFT!: "); @time IFFT! = FFTW.plan_ifft!(exfield.data, [1, 2]; flags=planner_flag)
-        print("DCT!:  "); @time DCT!  = FFTW.plan_r2r!(exfield.data, FFTW.REDFT10, 3; flags=planner_flag)
-        print("IDCT!: "); @time IDCT! = FFTW.plan_r2r!(exfield.data, FFTW.REDFT01, 3; flags=planner_flag)
-    else
-        FFT!  = FFTW.plan_fft!(exfield.data, [1, 2]; flags=planner_flag)
-        IFFT! = FFTW.plan_ifft!(exfield.data, [1, 2]; flags=planner_flag)
-        DCT!  = FFTW.plan_r2r!(exfield.data, FFTW.REDFT10, 3; flags=planner_flag)
-        IDCT! = FFTW.plan_r2r!(exfield.data, FFTW.REDFT01, 3; flags=planner_flag)
-    end
+    storage = zeros(Complex{Float64}, grid.Nx, grid.Ny, grid.Nz)
 
-    PoissonSolver{Array{eltype(g),1}}(kx², ky², kz², FFT!, DCT!, IFFT!, IDCT!)
+    for i in 1:Nx; kx²[i] = (2sin((i-1)*π/Nx)    / (Lx/Nx))^2; end
+    for j in 1:Ny; ky²[j] = (2sin((j-1)*π/Ny)    / (Ly/Ny))^2; end
+    for k in 1:Nz; kz²[k] = (2sin((k-1)*π/(2Nz)) / (Lz/Nz))^2; end
+
+    FFT!  = FFTW.plan_fft!(storage, [1, 2]; flags=planner_flag)
+    IFFT! = FFTW.plan_ifft!(storage, [1, 2]; flags=planner_flag)
+    DCT!  = FFTW.plan_r2r!(storage, FFTW.REDFT10, 3; flags=planner_flag)
+    IDCT! = FFTW.plan_r2r!(storage, FFTW.REDFT01, 3; flags=planner_flag)
+
+    PoissonSolverCPU{typeof(kx²), typeof(storage)}(kx², ky², kz², storage, FFT!, DCT!, IFFT!, IDCT!)
```

Comment:
Omit the manual typing of fields here to make concrete-ifying the fields of `PoissonSolverCPU` pain-free.

Reply:
type inference is your friend!

Reply:
Lol woops meant to give the :+1:

Reply:
Oh, the dreaded "emoji typo" --- bane of millennials everywhere!